### PR TITLE
feat: enable JWT authentication

### PR DESCRIPTION
**Pull request by Jesse0Michael**
Enable the `enable_jwt` api definition field and its supporting `jwt_*` api definition fields to begin configuring JWT authentication through the Tyk operator.

I did not add definitions to each of these fields, if there is an api definition reference that I could copy and paste the definitions from, please let me know. (instead of me guessing what would be good verbiage)

## Motivation and Context

We would like to try using JWT auth in the API's created by the operator

## Test Coverage For This Change

I tested this change locally with a tyk-operator build that included these changes and a local tyk gateway (community edition) using an ApiDefinition CRD that included these values
```
  enable_jwt: true
  jwt_signing_method: hmac
  jwt_source: dHlrMTIz
  jwt_identity_base_field: sub
  jwt_policy_field_name: pol
```
which created an api definition with the values
``` json
        "enable_jwt": true,
        "jwt_signing_method": "hmac",
        "jwt_source": "dHlrMTIz",
        "jwt_identity_base_field": "sub",
        "jwt_client_base_field": "",
        "jwt_policy_field_name": "pol",
        "jwt_default_policies": null,
        "jwt_issued_at_validation_skew": 0,
        "jwt_expires_at_validation_skew": 0,
        "jwt_not_before_validation_skew": 0,
        "jwt_skip_kid": false,
        "jwt_scope_to_policy_mapping": null,
        "jwt_scope_claim_name": "",
```

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). If PRing from your fork, don't come from your `master`!
- [x] Make sure you are making a pull request against our **`master` branch** (left side). Also, it would be best if you started *your change* off *our latest `master`*.
- [x] My change requires a change to the documentation.
  - [ ] If you've changed APIs, describe what needs to be updated in the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] Modules and vendor dependencies have been updated; run `go mod tidy && go mod vendor` in applicable directories.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] Check your code additions will not fail linting checks:
  - [x] `gofmt -s -w .`
  - [x] `go vet ./...`
  - [x] `golangci-lint run`

### CODE REVIEW

File: api/v1alpha1/apidefinition_types.go [hunk 1]
```diff
@@ -417,22 +417,24 @@ type APIDefinitionSpec struct {
...
-	//JWTPolicyFieldName         string                `json:"jwt_policy_field_name"`
-	//JWTDefaultPolicies         []string              `json:"jwt_default_policies"`
-	//JWTIssuedAtValidationSkew  uint64                `json:"jwt_issued_at_validation_skew"`
-	//JWTExpiresAtValidationSkew uint64                `json:"jwt_expires_at_validation_skew"`
-	//JWTNotBeforeValidationSkew uint64                `json:"jwt_not_before_validation_skew"`
-	//JWTSkipKid                 bool                  `json:"jwt_skip_kid"`
-	//JWTScopeToPolicyMapping    map[string]string     `json:"jwt_scope_to_policy_mapping"`
-	//JWTScopeClaimName          string                `json:"jwt_scope_claim_name"`
+	EnableCoProcessAuth        bool              `json:"enable_coprocess_auth,omitempty"`
+	JWTSigningMethod           string            `json:"jwt_signing_method,omitempty"`
```
Comment (asoorm) [inline api/v1alpha1/apidefinition_types.go:426]: Would be great to have kubebuilder validation here. something like:

```
// +kubebuilder:validation:Enum=rsa;hmac;ecdsa
```

File: api/v1alpha1/apidefinition_types.go [hunk 2]
```diff
@@ -417,22 +417,24 @@ type APIDefinitionSpec struct {
...
+	JWTSource                  string            `json:"jwt_source,omitempty"`
+	JWTIdentityBaseField       string            `json:"jwt_identity_base_field,omitempty"`
+	JWTClientIDBaseField       string            `json:"jwt_client_base_field,omitempty"`
+	JWTPolicyFieldName         string            `json:"jwt_policy_field_name,omitempty"`
+	JWTDefaultPolicies         []string          `json:"jwt_default_policies,omitempty"`
```
Comment (asoorm) [inline api/v1alpha1/apidefinition_types.go:431]: I need to think about how best we should handle this circular dependency.

The SecurityPolicy depends on the API Definition, and now If we enable JWT, the API Definition will depend on the Security Policy.

How to handle cleanup?
Is it possible to target the security policy by it's kubernetes name?

File: api/v1alpha1/apidefinition_types.go [hunk 3]
```diff
@@ -417,22 +417,24 @@ type APIDefinitionSpec struct {
...
+	JWTIssuedAtValidationSkew  uint64            `json:"jwt_issued_at_validation_skew,omitempty"`
+	JWTExpiresAtValidationSkew uint64            `json:"jwt_expires_at_validation_skew,omitempty"`
+	JWTNotBeforeValidationSkew uint64            `json:"jwt_not_before_validation_skew,omitempty"`
+	JWTSkipKid                 bool              `json:"jwt_skip_kid,omitempty"`
+	JWTScopeToPolicyMapping    map[string]string `json:"jwt_scope_to_policy_mapping,omitempty"`
```
Comment (asoorm) [inline api/v1alpha1/apidefinition_types.go:436]: Circular Dependency

**Comment by asoorm** [comment]
Thanks @Jesse0Michael - I'm thinking that we might merge this in order to just get JWT auth working and to parity with our existing API Definition object. However - my notes / comments are mainly things that we need to think about in order to offer a better experience.

**Review by gernest — APPROVED** [review]
we can merge as it is and I will cleanup for `0.5.0` release

**Comment by Jesse0Michael** [comment]
Yay! Thanks, @asoorm @gernest !

File: api/v1alpha1/apidefinition_types.go (hunk 2 above)
Comment (Jesse0Michael) [reply api/v1alpha1/apidefinition_types.go:431]: @asoorm  I didnt understand what you meant by this initially, but than ran into it when trying to get this to work locally
```
{
    "error": "key not authorized: no matching policy found"
}
```
Did you come up with a clever solution?

File: api/v1alpha1/apidefinition_types.go (hunk 3 above)
Comment (joshblakeley) [reply api/v1alpha1/apidefinition_types.go:436]: Have the scope mapping as a seperate mini CRD or secret?

**Pull request merged**